README.md
# Hawkscope

Cross-platform pluggable menu based launcher

## Building Hawkscope

### Prequisites
To build Hawkscope executable jar, you need:
 * JDK 5+
 * Maven 2.0.6+

### Adding SWT native library to local Maven repository

Hawkscope uses Eclipse SWT native libraries for each os / arch. Library
dependency is automatically selected via Maven profiles in pom.xml.
Current profiles are: win32, win64, linux32, linux64.

To install SWT native library to local Maven repository, find out your
system architecture (i.e. if you're on Windows XP, it is most probably 32 bit,
so your architecture would be win32), then install the appropriate jar from
Hawkscope project's /lib folder. Refer to lib/install.txt to find the
mvn install:install-file line that is suitable for you.

If you want to build Hawkscope and there is no profile for your operating system
and CPU architecture, you can create new Maven profile in pom.xml, download your
SWT implementation from http://www.eclipse.org/swt/ and install it. Please
feel free to contribute new profiles.

### Packaging Hawkscope + dependencies into executable JAR

To build an executable jar, simply run:

```console
$ mvn clean package
```

The jar will be in target/ folder, named: `hawkscope-(version)-(os)-(arch)-jar-with-dependencies.jar`

## Running Hawkscope

### Running Hawkscope from source folder
Running directly from source is good for debugging purposes. In console, execute:

```console
$ mvn clean compile exec:java -Dexec.mainClass=com.varaneckas.hawkscope.Launcher -e
```

### Running Hawkscope from executable JAR

Doubleclick the executable jar, or execute in console:

```console
$ java -jar hawkscope-<version>.jar
```


changelog.txt
$Id$
===================
HAWKSCOPE CHANGELOG
===================
http://hawkscope.googlecode.com

Changes in version 0.6.3 (2010-08-16)
-------------------------------------
* Fixed Mac OS X Leopard / Snow Leopard / 32bit / 64bit support
  http://code.google.com/p/hawkscope/issues/detail?id=122

Changes in version 0.6.2 (2009-06-21)
-------------------------------------
* Quick access list is now being saved properly on Windows
  http://code.google.com/p/hawkscope/issues/detail?id=109

Changes in version 0.6.1 (2009-06-07)
-------------------------------------
* System wide shortcut key for Windows and Linux
  http://code.google.com/p/hawkscope/issues/detail?id=35
* Possibility to rename Quick Access items
  http://code.google.com/p/hawkscope/issues/detail?id=15
* Disabled plugins are no longer class-loaded
  http://code.google.com/p/hawkscope/issues/detail?id=99
* Hawkscope icon is no longer invisible on dark backgrounds
  http://code.google.com/p/hawkscope/issues/detail?id=98
* Added a Donate button in About window
  http://code.google.com/p/hawkscope/issues/detail?id=100
  
Changes in version 0.6.0 (2009-04-04)
-------------------------------------
* Plugins can no longer break the rest of Hawkscope menu
  http://code.google.com/p/hawkscope/issues/detail?id=84
* Native icon themes for supported operating systems
  http://code.google.com/p/hawkscope/issues/detail?id=83
* Unified HTTP authentication mechanism for version checks and plugins that 
  use network connection 
  http://code.google.com/p/hawkscope/issues/detail?id=81  
* Ability to show/hide known file icons. Mac .app extension is always hidden.
  http://code.google.com/p/hawkscope/issues/detail?id=78
* Implemented checking for plugin updates
  http://code.google.com/p/hawkscope/issues/detail?id=53

Changes in version 0.5.4 (2009-03-21)
-------------------------------------
* Gmail Checker plugin
  http://code.google.com/p/hawkscope/issues/detail?id=76
* Open With plugin now saves/loads custom application paths well in Windows
  http://code.google.com/p/hawkscope/issues/detail?id=73
* Open With plugin was failing to open files with spaces on Linux 
  http://code.google.com/p/hawkscope/issues/detail?id=72
* Various enhancements of Execute plugin
  http://code.google.com/p/hawkscope/issues/detail?id=71
* Twitter plugin auto-recovery after network failures
  http://code.google.com/p/hawkscope/issues/detail?id=70    
* Improved keyboard navigation
  http://code.google.com/p/hawkscope/issues/detail?id=69
* No more problems with opening huge folders
  http://code.google.com/p/hawkscope/issues/detail?id=40  


Changes in version 0.5.3 (2009-03-13)
-------------------------------------
* Optimized memory consumption and overal performance
* Improved icon in Windows version
* Added the missing msvcr71.dll in Windows version
  http://code.google.com/p/hawkscope/issues/detail?id=67
* Fixed Network Settings proxy authentication checkbox behavior
  http://code.google.com/p/hawkscope/issues/detail?id=66
* Long menus are no longer splitted with "More..."
  http://code.google.com/p/hawkscope/issues/detail?id=64
* Periodic menu reload in background (with at least 3 minutes between reloads)
  http://code.google.com/p/hawkscope/issues/detail?id=62
* Execute plugin (available separately)
  http://code.google.com/p/hawkscope/issues/detail?id=60 

Changes in version 0.5.2 (2009-02-28)
-------------------------------------
* Twitter plugin (available separately)
  http://code.google.com/p/hawkscope/issues/detail?id=59
* Improved plugin support
  http://code.google.com/p/hawkscope/issues/detail?id=57
* Better password protection
  http://code.google.com/p/hawkscope/issues/detail?id=58

Changes in version 0.5.1 (2009-02-24)
-------------------------------------
* Linux GIJ 4.3 support
  http://code.google.com/p/hawkscope/issues/detail?id=54
  http://code.google.com/p/hawkscope/issues/detail?id=55
  http://code.google.com/p/hawkscope/issues/detail?id=56

Changes in version 0.5.0 (2009-02-20)
-------------------------------------
* Plugin System
  http://code.google.com/p/hawkscope/issues/detail?id=27
* Open With plugin
  http://code.google.com/p/hawkscope/issues/detail?id=42  
* A temporary workaround for Mac menubar icon blues
  http://code.google.com/p/hawkscope/issues/detail?id=33
* Friendlier GUI windows
  http://code.google.com/p/hawkscope/issues/detail?id=44   
* Interactive error report
  http://code.google.com/p/hawkscope/issues/detail?id=46 
* Help Window
  http://code.google.com/p/hawkscope/issues/detail?id=43

Changes in version 0.4.1 (2009-02-06)
-------------------------------------
* Separate versions for each operating system
  http://code.google.com/p/hawkscope/issues/detail?id=37
* Mac application contents are no longer shown
  http://code.google.com/p/hawkscope/issues/detail?id=38
* Fixed Mac OS X Dock Quit problem
  http://code.google.com/p/hawkscope/issues/detail?id=32

Changes in version 0.4.0 (2009-01-29)
-------------------------------------
* Made Hawkscope Java 5 compatible, now it runs on Mac OS X 10.4+
  http://code.google.com/p/hawkscope/issues/detail?id=9
* HTTP proxy authentication for update checks
  http://code.google.com/p/hawkscope/issues/detail?id=26
* Better handling of resources and smaller memory consumption
  http://code.google.com/p/hawkscope/issues/detail?id=28
* Added possibility to use operating system icons if they are available
  http://code.google.com/p/hawkscope/issues/detail?id=24 
* Swing GUI implementation is now discontinued and removed
  http://code.google.com/p/hawkscope/issues/detail?id=22
* Added possibility to override configuration parameters with args line
  http://code.google.com/p/hawkscope/issues/detail?id=23
* Added possibility to blacklist folders and drives
  http://code.google.com/p/hawkscope/issues/detail?id=16 
* Settings Window

src/main/java/com/varaneckas/hawkscope/util/OSUtils.java
/*
 * Copyright (c) 2008-2009 Tomas Varaneckas
 * http://www.varaneckas.com
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.varaneckas.hawkscope.util;

import java.io.File;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import javax.swing.filechooser.FileSystemView;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.eclipse.swt.widgets.Text;

/**
 * Operating System specific utilities.
 * 
 * Used for places where different OS acts differently.
 * 
 * @author Tomas Varaneckas
 * @version $Id$
 */
public abstract class OSUtils {
	
    /**
     * Logger
     */
	private static final Log log = LogFactory.getLog(OSUtils.class);

	/**
	 * Supported Operating Systems
	 * 
	 * @author Tomas Varaneckas
	 * @version $Id$
	 */
	public enum OS {
		MAC, WIN, UNIX, UNKNOWN
	}
	
	/**
	 * Current Operating System
	 */
	public static final OS CURRENT_OS = getCurrentOS();

	/**
	 * FileSystemView (don't use with Mac)
	 */
	private static final FileSystemView fsw = getFileSystemView();

	/**
	 * Gets the current operating system
	 * 
	 * @return
	 */
	private static OS getCurrentOS() {
		final String os = System.getProperty("os.name", "unknown").toLowerCase();
		if (os.startsWith("win")) {
			return OS.WIN;
		}
		if (os.startsWith("mac")) {
			return OS.MAC;
		}
		if (os.startsWith("linux") || os.startsWith("unix")
				|| os.startsWith("bsd")) {
			return OS.UNIX;
		}
		return OS.UNKNOWN;
	}

	/**
	 * Gets FileSystemView for non-macs
	 * 
	 * @return
	 */
	private static FileSystemView getFileSystemView() {
	    if (!CURRENT_OS.equals(OS.MAC)) {
	        return FileSystemView.getFileSystemView();
	    }
        return null;
    }

    /**
	 * Gets the tray icon size
	 * 
	 * @return
	 */
	public static int getTrayIconSize() {
		//experimental
		if (System.getProperty("java.version").compareTo("1.6") >= 0) {
			try {
				log.debug("Java > 1.6, trying java.awt.SystemTray");
				final Class<?> systemTrayClass = Class.forName(
				        "java.awt.SystemTray");
				Method m = systemTrayClass.getMethod("getSystemTray", 
				        new Class[] {});
				final Object systemTray = m.invoke(systemTrayClass, 
				        new Object[] {});
				m = systemTray.getClass().getMethod("getTrayIconSize", 
				        new Class[] {});
				final Object size = m.invoke(systemTray, new Object[] {});
				m = null;
				return ((java.awt.Dimension) size).height;
			} catch (final Exception e) {
				log.warn("Failed calling java.awt.SystemTray object", e);
			}
		}
		if (CURRENT_OS.equals(OS.WIN)) {
			return 16;
		}
		if (CURRENT_OS.equals(OS.UNIX)) {
			return 24;
		}
		if (CURRENT_OS.equals(OS.MAC)) {
			return 16;
		}
		return 16;
	}

	/**
	 * Gets the system tray (menubar, panel, whatever) size
	 * 
	 * @return
	 */
	public static int getTraySize() {
		if (CURRENT_OS.equals(OS.MAC)) {
			return 26;
		}
		return getTrayIconSize();
	}

	/**
	 * Gets file display name according to local OS
	 * 
	 * @param file
	 * @return
	 */
	public static String getSystemDisplayName(final File file) {
		if (!CURRENT_OS.equals(OS.MAC)) {
			return fsw.getSystemDisplayName(file);
		}
		return null;
	}

	/**
	 * Tells if {@link File} is a floppy drive
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isFloppyDrive(final File file) {
		if (CURRENT_OS.equals(OS.WIN)) {
			return fsw.isFloppyDrive(file);
		}
		return false;
	}

	/**
	 * Tells if {@link File} is a file system root
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isFileSystemRoot(final File file) {
		switch (CURRENT_OS) {
		case MAC: {
		    if (file.getAbsolutePath().equals("/") || file.getAbsolutePath()
		            .matches("^/Volumes/[^/]+$")) {
		        return true;
		    }
		    return false;
		}
		case UNIX: {
		    if (file.getAbsolutePath().equals("/") 
		            || file.getAbsolutePath().matches("^/(media|mnt)/[^/]+$")) {
		        return true;
		    }
		    return false;
		}
		default: 
		    return fsw.isFileSystemRoot(file);
		}
	}

	/**
	 * Gets the File System Roots
	 * 
	 * @return
	 */
	public static List<File> getFileSystemRoots() {
		final List<File> roots = new LinkedList<File>();
		switch (CURRENT_OS) {
		case MAC: {
			final File[] volumes = new File("/Volumes").listFiles();
			if (volumes.length == 0) {
				//something's wrong..
				roots.add(new File("/"));
			} else {
				for (final File f : volumes) {
					if (f.canRead() && !isMacApp(f)) {
					    roots.add(f);
					}
				}
			}
			return roots;
		}
		case UNIX:
			roots.add(new File("/"));
			//Ubuntu
			File media = new File("/media");
			if (!media.exists()) {
			    //Old School Unix fallback
			    media = new File("/mnt");
			} 
			if (media.exists()) {
    			for (final File f : media.listFiles()) {
    				if (f.isDirectory() && f.canRead()) {
    					roots.add(f);
    				}
    			}
			}
			return roots;
		default: {
			return Arrays.asList(File.listRoots());
		}
		}
	}
	
	/**
	 * Tells if {@link File} is an optical drive
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isOpticalDrive(final File file) {
	    if (!file.isDirectory()) {
	        return false;
	    }
	    String name = null;
	    if (CURRENT_OS.equals(OS.WIN)) {
	        name = fsw.getSystemDisplayName(file)
	                .toLowerCase();
	    }
	    if (name == null) {
	        name = file.getName().toLowerCase();
	    }
	    return name.matches(".*(cdrom|dvd).*");
	}
	
	/**
	 * Tells if file is a network drive
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isNetworkDrive(final File file) {
	    if (!file.isDirectory()) {
	        return false;
	    }
	    if (isMacApp(file)) {
	        return false;
	    } 
	    if (CURRENT_OS.equals(OS.WIN)) {
	        if (getSystemDisplayName(file).contains(" on ")) {
	            return true;
	        }
	        return fsw.isComputerNode(file);
	    } 
	    return file.getAbsolutePath().toLowerCase()
	        .matches(".*(server|network|remote).*");
	}

	/**
	 * Tells if file is a removable drive
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isRemovableDrive(final File file) {
	    if (!file.isDirectory()) {
	        return false;
	    }
	    if (isMacApp(file)) {
	        return false;
	    }
        String name = null;
        if (CURRENT_OS.equals(OS.WIN)) {
            name = fsw.getSystemDisplayName(file).toLowerCase();
        }
        if (name == null) {
            name = file.getName().toLowerCase();
        }
        return name.matches(".*(usb|flash|removable).*");
	}
	
	/**
	 * Tells if file is executable
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isExecutable(final File file) {
	    final String filePath = file.getAbsolutePath().toLowerCase();
	    switch (CURRENT_OS) {
        case WIN:
            if (filePath.endsWith(".exe") 
                    || filePath.endsWith(".bat")
                    || filePath.endsWith(".msi")) {
                
                return true;
            } else {
                return false;
            }
        default:
            if (System.getProperty("java.version").compareTo("1.6") >= 0) {
                //we can check if it's executable
                try {
                    Method m = file.getClass().getMethod("canExecute", 
                            new Class[] {});
                    if (m != null) { 
                        final boolean result = (Boolean) m.invoke(file, 
                                new Object[] {}); 
                        m = null;
                        return result;
                    }
                } catch (final Exception e) {
                    log.warn("Failed dynamically calling File.canExecute", e);
                    return false;
                }
            } 
            return filePath.contains("/bin/");
        }
	}
	
	/**
	 * Tells if file is a Mac application
	 * 
	 * @param file
	 * @return
	 */
	public static boolean isMacApp(final File file) {
	    if (CURRENT_OS.equals(OS.MAC) && file != null) {
	        return file.isDirectory() 
	            && file.getName().toLowerCase().endsWith(".app");
	    } 
	    return false;
	}

	/**
	 * Executes app with parameters
	 * 
	 * @param app command
	 * @param params parameters
	 * @return needs further execution? (when failed, returns true, otherwise false)
	 */
	public static boolean exec(String app, String params) {
		try {
		    final List<String> cmd = new ArrayList<String>();
			switch (CURRENT_OS) {
			case MAC:
				if (app.toLowerCase().endsWith(".app")) {
				    cmd.add("open");
				    cmd.add("-a");
				    cmd.add(app);
				    cmd.add(params);
				    break;
				} 
			default:
			    cmd.add(app);
			    cmd.add(params);
			}
			if (log.isDebugEnabled()) {
				log.debug("Executing: " + cmd);
			}
			String[] cmdArray = new String[] {};
			cmdArray = cmd.toArray(cmdArray);
			Runtime.getRuntime().exec(cmdArray);
			cmdArray = null;
			return false;
		} catch (final Exception e) {
			log.warn("Failed executing app " + app + " with params: " 
			        + params, e);
		}
		return true;
	}
	
	/**
	 * Sleep for some milliseconds. No exceptions.
	 * 
	 * @param millis 
	 */
	public static void sleep(final long millis) {
		try {
			Thread.sleep(millis);
		} catch (InterruptedException e) {
			log.warn("Thread sleep interrupted", e);
		}
	}

	/**
	 * Gets SWT {@link Text} top offset adjustment. Apply manually.
	 * 
	 * @return
	 */
    public static int getTextTopOffsetAdjust() {
        switch (CURRENT_OS) {
        case WIN: 
            return -2;
        case MAC:
        	return -7;
        default:
            return -5;
        }
    }

    /**
     * Gets platform friendly fixed font name
     * 
     * @return
     */
	public static String getFixedFontName() {
		switch (CURRENT_OS) {
		case MAC:
			return "Monaco";
		case UNIX:
		    return "Monospace";
		default:
			return "Courier New";
		}
	}
}
